packages/react-dom/src/__tests__/ReactInputSelection-test.js

```diff
@@ -0,0 +1,254 @@
+/**
+ * Copyright (c) 2013-present, Facebook, Inc.
+ *
+ * This source code is licensed under the MIT license found in the
+ * LICENSE file in the root directory of this source tree.
+ *
+ * @emails react-core
+ */
+
+'use strict';
+
+var React = require('react');
+var ReactDOM = require('react-dom');
+var ReactTestUtils = require('react-dom/test-utils');
+var ReactInputSelection = require('../client/ReactInputSelection');
+
+describe('ReactInputSelection', () => {
+  var textValue = 'the text contents';
+  var createAndMountElement = (type, props, children) => {
+    var element = React.createElement(type, props, children);
+    var instance = ReactTestUtils.renderIntoDocument(element);
+    return ReactDOM.findDOMNode(instance);
+  };
+  var makeGetSelection = (win = window) => () => ({
+    anchorNode: win.document.activeElement,
+    focusNode: win.document.activeElement,
+    anchorOffset:
+      win.document.activeElement && win.document.activeElement.selectionStart,
+    focusOffset:
+      win.document.activeElement && win.document.activeElement.selectionEnd,
+  });
+
+  describe('hasSelectionCapabilities', () => {
+    it('returns true for textareas', () => {
+      var textarea = document.createElement('textarea');
+      expect(ReactInputSelection.hasSelectionCapabilities(textarea)).toBe(true);
+    });
+
+    it('returns true for text inputs', () => {
+      var inputText = document.createElement('input');
+      var inputReadOnly = document.createElement('input');
+      inputReadOnly.readOnly = 'true';
+      var inputNumber = document.createElement('input');
+      inputNumber.type = 'number';
+      var inputEmail = document.createElement('input');
+      inputEmail.type = 'email';
+      var inputPassword = document.createElement('input');
+      inputPassword.type = 'password';
+      var inputHidden = document.createElement('input');
+      inputHidden.type = 'hidden';
+
+      expect(ReactInputSelection.hasSelectionCapabilities(inputText)).toBe(
+        true,
+      );
+      expect(ReactInputSelection.hasSelectionCapabilities(inputReadOnly)).toBe(
+        true,
+      );
+      expect(ReactInputSelection.hasSelectionCapabilities(inputNumber)).toBe(
+        false,
+      );
+      expect(ReactInputSelection.hasSelectionCapabilities(inputEmail)).toBe(
+        false,
+      );
+      expect(ReactInputSelection.hasSelectionCapabilities(inputPassword)).toBe(
+        false,
+      );
+      expect(ReactInputSelection.hasSelectionCapabilities(inputHidden)).toBe(
+        false,
+      );
+    });
```

Comment:
These feel like they could be separate tests. 

Also: Why do `email`, `password`, and `number` return false? Is this because they don't support the text selection api?

Reply:
That’s a great question! That’s what the existing function enforces, so I figured I’d capture that behavior in the tests:

```js
  return (
    nodeName &&
    ((nodeName === 'input' && elem.type === 'text') ||
      nodeName === 'textarea' ||
      elem.contentEditable === 'true')
  );
```

But I’m not sure it’s desired, because those input types can definitely have selections. The `hasSelectionCapabilities` check could be changed to something like:
```js
  return (
    nodeName &&
    ((nodeName === 'input' && ['text', 'email', 'number', 'password'].includes(elem.type)) ||
      nodeName === 'textarea' ||
      elem.contentEditable === 'true')
  );
```